### DISCUSSION TITLE
tags and filters

### DISCUSSION THREAD

**Discussion by Waldorf3** [docker-mailserver/docker-mailserver#3797] [Ideas]
Possibly a dumb or irrelevant question, but is there any plans or considerations for adding tags and filters similar to what is used in Gmail?  

I really want to get rid of Gmail but being able to tag and filter incoming mail is really a killer feature. I realize something like this may require a unified frontend, but what do I know.  Just something I really would like to see if possible.

**Comment by casperklein** [comment]
Not sure if that fits your needs, but you could take a look at [roundcube](https://github.com/roundcube/roundcubemail) as a web UI. There is also a docker variant, which can easily be used with DMS.
For tags and filters you can take a look sieve. There is also a plugin for roundcube afaik called managesieve.

**Comment by Waldorf3** [comment]
I've seen sieve mentioned but never thought much about what it was used for, now I understand.  So just a matter of reading up on how to use it.  I found what seems to be a pretty good tutorial [here ](https://p5r.uk/blog/2011/sieve-tutorial.html)I'll get started with. Thanks.

**Comment by Waldorf3** [comment]
Maybe just a quick (I hope) clarification to [Email Filtering with Sieve](https://docker-mailserver.github.io/docker-mailserver/latest/config/advanced/mail-sieve/) page.   It says

`Global filters are applied to EVERY incoming mail for EVERY email address. To specify a global Sieve filter provide a docker-data/dms/config/before.dovecot.sieve`

I don't find any docker-data/dms/config tree anywhere in the container.  The closest is config bound to /tmp/docker-mailserver - but having it there does not seem to work.  Where exactly in the container is before.dovecot.sieve supposed to be placed for it to be picked up?  

the before.dovecot.sieve is very simple:

```
require ["fileinto", "reject"];

if address :is ["from", "sender"] "<email@address>"
{
    reject "Go away, please";
}

```

**Comment by casperklein** [comment]
> I don't find any docker-data/dms/config tree anywhere in the container. 

That's the host directory, that is mounted into the container:

https://github.com/docker-mailserver/docker-mailserver/blob/3cbcdb2d6517b6c61818afb49c0eb58ee82b6202/compose.yaml#L21

> but having it there does not seem to work

I can't help you with that, I've never used sieve. You may open a bug report.

PS: Make sure to correctly restart DMS, after you added the file.

**Comment by Waldorf3** [comment]
After some trial and error, I found that if I leave out the <> characters it works, i.e.  `if address :is ["from", "sender"] "email@address"` - maybe they need to be escaped?

Would it not be enough to restart the imap server? (docker exec mailserver supervisorctl restart dovecot)  Restarting the whole DMC seems to require stopping and starting it and takes forever...

**Reply by casperklein** [reply]
> After some trial and error, I found that if I leave out the <> characters it works

Why did you use them in the first place?

> Would it not be enough to restart the imap server? 

Unfortunately not. `docker-data/dms/config/before.dovecot.sieve` is only handled during DMS startup.

**Reply by ourhelpdesk** [reply]
I read somewhere here that without the <> you can risk partical matches, i.e. "mail@address" would also match "someemail@address". Anyway, no big deal.

**Reply by polarathene** [reply]
> Would it not be enough to restart the imap server?

`check-for-changes.sh` just needs to add support to monitor the config volume file to copy it over when it's updated. Someone would need to add the path. Fairly simple contribution.

If you edit the sieve file at the internal location it gets copied to, you should be able to just run `dovecot reload` if needed and it'll refresh it's config.

Much faster that way for experimenting (_`check-for-changes.sh` could be nicer but often more naive and can restart/reload other services as part of it's process_)..

**Reply by polarathene** [reply]
> I read somewhere here that without the <> you can risk partical matches, i.e. "mail@address" would also match "someemail@address".

That should depend on the match type used, default is exact match with `:is`. If you use `:contains` then it'll allow partial matches.

**Reply by kour1er** [reply]
@casperklein 

> `before.dovecot.sieve` is only handled during DMS startup

I'm probably missing something obvious, but I'm finding before.dovecot.sieve in config is only updated in /usr/lib/dovecot/sieve-global/before/50-before.dovecot.sieve when i --force-recreate the container. Is that the intended behaviour? I'm not seeing any obvious errors in the debug log

**Reply by casperklein** [reply]
Yes. For now, you should always user `docker compose down/up` or `--force-recreate` to ensure a fresh container at startup. Traditional container restarts are not fully supported yet.

**Reply by kour1er** [reply]
Ah - good to know I'm not being totally dumb! Thank you for confirming
